### DISCUSSION TITLE
[Fast-Data] Create primary key with fields not in Kafka message key

### DISCUSSION THREAD

**Discussion by gino-ats** [mia-platform/community#516] [Q&A] [answered]
### What product are you having troubles with?

Fast Data

### What Console version are you using?

v12.0.2

### Description

We are trying to create a Projection with a primary key that involves fields non included in the Kafka key of the message received from Debezium but we receive “one of primary key fields defined in projection config has not been found in the record key”.  Is there a way to solve it ? 

### Actual Outcome

Currently the console allow to configure this but when we deploy it we receive the error:  “one of primary key fields defined in projection config has not been found in the record key”

**Comment by flower-of-the-bridges** [comment]
Hi @gino-ats

The primary key fields must match the fields in your Kafka key: this is since ingestion services such as Real-Time Updater or Projection Storer must preserve the order of messages in the ingestion topic. The console allows it since the configuration can't know which kind of messages will be sent to the event streaming platform.

I strongly suggest changing the CDC configuration, to match the primary keys you have in your configuration, so you can preserve order of messages and avoid side effects.

Bests
Giovanni

**Reply by gino-ats** [reply]
Ok thanks, it makes sense.  But let me deep dive the issue. The problem comes from the absence of the key fields into the body of message, because the original source is a kind of log table without primary key.  So a "surrogate" key was added but it is not present in the message body (because they don't want to change the table).  Now, when we use this surrogate key as primary key, the key field is not added to the mongodb collection data that remains with an empty key.  So, when the second message is received it raise a "duplicate key" exception.  Have you some idea how to solve it ?

**Reply by flower-of-the-bridges** [reply]
Hi @gino-ats 

If I understood correctly: this surrogate key is only defined in the kafka message key, but not on the kafka message value. Is that correct?

**Reply by gino-ats** [reply]
yes, right !

**Reply by flower-of-the-bridges** [reply, marked as answer]
Ok got it, thanks for the clarification.

I think the solution to this is to define a custom message adapter that injects the missing key value into the value object:

```js
'use strict'

function messageAdapter(message, primaryKeys, logger) {
    const { value, key: keyAsString } = message

    const keyObject = JSON.parse(keyAsString)
    const payload = JSON.parse(value.toString())
    
    logger.trace('received message in custom message adapter')

   // 💡 Here I add the key value to the payload
   payload['my-custom-key'] = keyObject['my-custom-key']

    return {
        operation: "this value you have to populate accordingly to your cdc configuration",
        key: keyObject,
        before: null,
        after: payload
    }
}

// the following code allows to use the same custom function
// in both the Projection Storer and the Real-Time Updater
try {
    // export function for Real-Time Updater
    module.exports = messageAdapter
} catch (error) {
    // ignore error when importing the custom function in the Projection Storer
    // since it exploits the function name
}

```

Keep in mind that this function will be performed on all messages from all projections in the system of record, so:

- either you have a dedicated system of record for this table
or
- the payload contains the name of the table and you can use a condition to add the value accordingly

For further details about custom message adapters [you can refer to this section of the doc](https://docs.mia-platform.eu/docs/fast_data/configuration/realtime-updater/configuration-files#custom). 

Let me know if this can be a valid solution for your case

Bests 
Giovanni

**Reply by gino-ats** [reply]
Write our custom adapter is a way that we would like to avoid due to complexity (we have to study the Debezium message in deep, handle snapshot, tombstone, ...) and plan a test list for them. Probably, some kind of "fields mapper" on top of current Debezium adapter could be a feature proposal.  Anyway, if there are no other options, we will investigate it. Thanks
